File: utils/startup_check.py
"""
Startup checks untuk memastikan semua dependencies dan config ready
"""
import sys
import os
from pathlib import Path

def check_dependencies():
    """Check if all required dependencies are installed"""
    missing = []
    
    required = {
        'discord': 'discord.py',
        'dotenv': 'python-dotenv'
    }
    
    optional = {
        'yt_dlp': 'yt-dlp (untuk music)',
        'spotipy': 'spotipy (untuk Spotify API)',
        'psutil': 'psutil (untuk resource monitoring)'
    }
    
    print("🔍 Checking dependencies...")
    
    # Check required
    for module, name in required.items():
        try:
            __import__(module)
            print(f"  ✅ {name}")
        except ImportError:
            print(f"  ❌ {name} - REQUIRED!")
            missing.append(name)
    
    # Check optional
    for module, name in optional.items():
        try:
            __import__(module)
            print(f"  ✅ {name} (optional)")
        except ImportError:
            print(f"  ⚠️  {name} (optional) - tidak terinstall")
    
    if missing:
        print(f"\n❌ Missing required dependencies: {', '.join(missing)}")
        print("Install dengan: pip install -r requirements.txt")
        return False
    
    return True

def check_config():
    """Check config file"""
    from utils.config_validator import validate_config, print_validation_results
    
    print("\n🔍 Checking config file...")
    
    if not os.path.exists('config.json'):
        print("  ❌ config.json tidak ditemukan!")
        print("  💡 Copy config.json.example ke config.json dan edit")
        return False
    
    valid, errors = validate_config()
    print_validation_results(valid, errors)
    
    return valid

def check_directories():
    """Check and create required directories"""
    print("\n🔍 Checking directories...")
    
    dirs = ['logs', 'backups']
    
    for dir_name in dirs:
        dir_path = Path(dir_name)
        if not dir_path.exists():
            dir_path.mkdir(exist_ok=True)
            print(f"  ✅ Created directory: {dir_name}/")
        else:
            print(f"  ✅ Directory exists: {dir_name}/")
    
    return True

def check_permissions():
    """Check file permissions"""
    print("\n🔍 Checking permissions...")
    
    # Check if we can write to current directory
    try:
        test_file = Path('.write_test')
        test_file.touch()
        test_file.unlink()
        print("  ✅ Write permission OK")
        return True
    except Exception as e:
        print(f"  ❌ Cannot write to current directory: {e}")
        return False

def run_all_checks():
    """Run all startup checks"""
    print("=" * 50)
    print("🛡️  SATPAM BOT - STARTUP CHECKS")
    print("=" * 50)
    print()
    
    checks = [
        ("Dependencies", check_dependencies),
        ("Config", check_config),
        ("Directories", check_directories),
        ("Permissions", check_permissions)
    ]
    
    results = []
    for name, check_func in checks:
        try:
            result = check_func()
            results.append((name, result))
        except Exception as e:
            print(f"  ❌ Error in {name} check: {e}")
            results.append((name, False))
    
    print("\n" + "=" * 50)
    print("📊 CHECK RESULTS")
    print("=" * 50)
    
    all_passed = True
    for name, result in results:
        status = "✅ PASS" if result else "❌ FAIL"
        print(f"  {status} - {name}")
        if not result:
            all_passed = False
    
    print()
    
    if all_passed:
        print("✅ All checks passed! Bot siap untuk dijalankan.")
        return True
    else:
        print("❌ Beberapa checks gagal. Silakan perbaiki sebelum menjalankan bot.")
        return False

if __name__ == "__main__":
    success = run_all_checks()
    sys.exit(0 if success else 1)

File: utils/startupCheck.js
import fs from 'fs';
import { loadConfig, getBotTokens, getConfigPath } from './config.js';

/**
 * Check dependencies
 */
export function checkDependencies() {
    console.log('🔍 Checking dependencies...');
    
    // Check if node_modules exists
    if (!fs.existsSync('node_modules/discord.js')) {
        console.log('  ❌ discord.js - REQUIRED!');
        console.log('  💡 Install dengan: npm install');
        return false;
    }
    console.log('  ✅ discord.js');

    if (!fs.existsSync('node_modules/dotenv')) {
        console.log('  ❌ dotenv - REQUIRED!');
        console.log('  💡 Install dengan: npm install');
        return false;
    }
    console.log('  ✅ dotenv');

    if (!fs.existsSync('node_modules/@discordjs/voice')) {
        console.log('  ❌ @discordjs/voice - REQUIRED!');
        console.log('  💡 Install dengan: npm install @discordjs/voice');
        return false;
    }
    console.log('  ✅ @discordjs/voice');

    // Check for encryption library (libsodium-wrappers)
    if (!fs.existsSync('node_modules/libsodium-wrappers')) {
        console.log('  ⚠️  libsodium-wrappers - REQUIRED untuk voice encryption');
        console.log('  💡 Install dengan: npm install libsodium-wrappers');
        return false;
    }
    console.log('  ✅ libsodium-wrappers (voice encryption)');

    return true;
}

/**
 * Check config file
 */
export function checkConfig() {
    console.log('\n🔍 Checking config file...');
    
    try {
        const config = loadConfig(true); // Force reload to ensure latest config
        if (!config) {
            console.log('  ❌ config.json tidak ditemukan!');
            console.log('  💡 Copy config.json.example ke config.json dan edit');
            console.log('  💡 Lokasi yang dicari:');
            console.log('     - ./config.json (root directory)');
            console.log('     - ./config/config.json');
            console.log('  💡 Contoh command:');
            console.log('     cp config/config.json.example config.json');
            console.log('     # atau jika di root:');
            console.log('     cp config.json.example config.json');
            return false;
        }

        // Validate config structure
        if (!config.bot_tokens || !Array.isArray(config.bot_tokens)) {
            console.log('  ❌ config.json tidak memiliki field "bot_tokens" yang valid!');
            console.log('  💡 Format harus: {"bot_tokens": ["token1", "token2", ...]}');
            return false;
        }

        const tokens = getBotTokens();
        if (tokens.length === 0) {
            console.log('  ❌ Tidak ada bot token yang valid!');
            console.log('  💡 Pastikan token tidak kosong dan bukan placeholder');
            console.log('  💡 Token harus minimal 50 karakter');
            return false;
        }

        const configPath = getConfigPath();
        console.log(`  ✅ config.json ditemukan: ${configPath || 'unknown'}`);
        console.log(`  ✅ ${tokens.length} bot token(s) valid`);
        
        if (config.bot_tokens && config.bot_tokens.length > tokens.length) {
            const placeholderCount = config.bot_tokens.length - tokens.length;
            console.log(`  ⚠️  ${placeholderCount} token(s) masih placeholder/kosong (akan diabaikan)`);
        }

        // Check optional configs
        if (config.idle_voice_channel_id) {
            const idleId = String(config.idle_voice_channel_id);
            console.log(`  ✅ Idle channel ID: ${idleId}`);
            console.log(`     💡 Tip: Gunakan string (e.g. "1451645891600453652") untuk menghindari precision issues`);
        } else {
            console.log(`  ℹ️  Idle channel ID tidak di-set (opsional)`);
        }

        if (config.music_enabled_bot) {
            console.log(`  ✅ Music enabled bot: #${config.music_enabled_bot}`);
        }

        if (config.spotify) {
            const spotify = config.spotify;
            if (spotify.client_id && spotify.client_secret && 
                !spotify.client_id.includes('your_') && !spotify.client_secret.includes('your_')) {
                console.log(`  ✅ Spotify credentials ditemukan`);
            } else {
                console.log(`  ℹ️  Spotify credentials tidak di-set (opsional)`);
            }
        }

        return true;
    } catch (error) {
        console.log(`  ❌ Error checking config: ${error.message}`);
        return false;
    }
}

/**
 * Check directories
 */
export function checkDirectories() {
    console.log('\n🔍 Checking directories...');
    
    const dirs = ['logs', 'backups'];
    let allOk = true;

    for (const dir of dirs) {
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
            console.log(`  ✅ Created directory: ${dir}/`);
        } else {
            console.log(`  ✅ Directory exists: ${dir}/`);
        }
    }

    return true;
}

/**
 * Run all checks
 */
export function runAllChecks() {
    console.log('='.repeat(50));
    console.log('🛡️  SATPAM BOT - STARTUP CHECKS');
    console.log('='.repeat(50));
    console.log();

    const checks = [
        { name: 'Dependencies', func: checkDependencies },
        { name: 'Config', func: checkConfig },
        { name: 'Directories', func: checkDirectories }
    ];

    const results = [];
    for (const check of checks) {
        try {
            const result = check.func();
            results.push({ name: check.name, result });
        } catch (error) {
            console.error(`  ❌ Error in ${check.name} check:`, error);
            results.push({ name: check.name, result: false });
        }
    }

    console.log('\n' + '='.repeat(50));
    console.log('📊 CHECK RESULTS');
    console.log('='.repeat(50));

    let allPassed = true;
    for (const { name, result } of results) {
        const status = result ? '✅ PASS' : '❌ FAIL';
        console.log(`  ${status} - ${name}`);
        if (!result) allPassed = false;
    }

    console.log();

    if (allPassed) {
        console.log('✅ All checks passed! Bot siap untuk dijalankan.');
        return true;
    } else {
        console.log('❌ Beberapa checks gagal. Silakan perbaiki sebelum menjalankan bot.');
        return false;
    }
}

